File: src/skeleton/fill.ts
import {
  distanceToProfile,
  envelopeMaxRadius,
  envelopeProfile,
  envelopeRadiusAt,
  type Envelope,
} from "../envelope";
import type { Skeleton } from "./colonize";
import { DEFAULT_SHED } from "./shed";

/* ------------------------------------------------------------------ *
 * FILL, MEASURED ON THE WOOD THE CALLER CLASSIFIES
 *
 * `terminal` names the nodes under measurement. The old pass supplies
 * appended nodes without children; the branch pass will supply its twig
 * marks. A zero-order skeleton therefore has no tested fine wood.
 *
 * Shell cells are equal-volume voxels whose CENTRES lie inside the
 * crown and pass shedding's predicate. The world-origin grid uses
 * `cellSize` as a fraction of envelope height (0.02 at rest). A cell
 * counts once if it holds at least one terminal position in the shell.
 * The fraction approximates occupied shell volume at this resolution;
 * it does not measure wood volume or credit a segment crossing a cell.
 * Surface-straddling cells are classified by centre on both sides of
 * the ratio. Out-of-crown terminals cannot inflate shell occupancy.
 *
 * Shell depth is DEFAULT_SHED.shellDepth * envelopeMaxRadius in METRES,
 * exactly as in shed.ts. The radial-slack shortcut and distanceToProfile
 * predicate are the shedder's too; the finite crown domain additionally
 * excludes the outside points that shedding conservatively retains.
 *
 * Clustering uses Euclidean distance in metres, inclusive at the bound.
 * Colonization tips have no children WITHIN [0, crossover), even if the
 * local pass later appended children. Each terminal counts once, however
 * many tips are nearby. No terminals, no tips, invalid inputs, or a grid
 * too large to measure are explicitly untested, never an invented zero.
 * ------------------------------------------------------------------ */

export type FillMeasurement =
  | { tested: true; fraction: number; numerator: number; denominator: number }
  | { tested: false; reason: string };

type Untested = Extract<FillMeasurement, { tested: false }>;
const untested = (reason: string): Untested => ({ tested: false, reason });
const fraction = (numerator: number, denominator: number): FillMeasurement =>
  ({ tested: true, fraction: numerator / denominator, numerator, denominator });

function terminals(
  skeleton: Skeleton,
  terminal: Uint8Array,
): number[] | Untested {
  if (terminal.length !== skeleton.nodes.length) {
    return untested("terminal-size-mismatch");
  }
  const indices: number[] = [];
  for (let i = 0; i < terminal.length; i += 1) {
    if (terminal[i] === 0) continue;
    const { x, y, z } = skeleton.nodes[i].position;
    if (![x, y, z].every(Number.isFinite)) return untested("non-finite-terminal");
    indices.push(i);
  }
  return indices.length ? indices : untested("no-terminals");
}

/** Fraction of shell voxels containing caller-classified terminal wood.
 *  `cellSize` is a height fraction, held constant across comparisons.
 *  More than two million candidate cells is reported as untested; the
 *  function never silently coarsens the caller's measurement resolution. */
export function shellOccupancy(
  skeleton: Skeleton,
  terminal: Uint8Array,
  envelope: Envelope,
  cellSize = 0.02,
): FillMeasurement {
  const selected = terminals(skeleton, terminal);
  if (!Array.isArray(selected)) return selected;
  if (!(cellSize > 0) || !Number.isFinite(cellSize)) {
    return untested("invalid-cell-size");
  }
  const { height, crownBase, spread, fullness, shoulder } = envelope;
  if (![height, crownBase, spread, fullness, shoulder].every(Number.isFinite)
    || height <= 0 || crownBase < 0 || crownBase >= 1 || spread <= 0 || shoulder <= 0) {
    return untested("invalid-envelope");
  }
  const size = cellSize * height;
  const maxRadius = envelopeMaxRadius(envelope);
  const base = height * crownBase;
  if (!Number.isFinite(size) || size <= 0 || !Number.isFinite(maxRadius)) {
    return untested("invalid-envelope");
  }
  const lo = Math.floor(-maxRadius / size);
  const hi = Math.ceil(maxRadius / size);
  const bottom = Math.floor(base / size);
  const top = Math.ceil(height / size);
  const cells = (hi - lo) ** 2 * (top - bottom);
  if (![lo, hi, bottom, top, cells].every(Number.isSafeInteger) || cells > 2_000_000) {
    return untested("grid-too-large");
  }

  const shell = DEFAULT_SHED.shellDepth * maxRadius;
  const profile = envelopeProfile(envelope);
  const inShell = (x: number, y: number, z: number): boolean => {
    if (y < base || y > height) return false;
    const r = Math.hypot(x, z);
    const slack = envelopeRadiusAt(envelope, y) - r;
    if (slack < 0) return false;
    return !(slack > shell) || !(distanceToProfile(profile, r, y) > shell);
  };
  const key = (x: number, y: number, z: number): string => `${x},${y},${z}`;
  const occupied = new Set<string>();
  for (const i of selected) {
    const { x, y, z } = skeleton.nodes[i].position;
    if (inShell(x, y, z)) {
      occupied.add(key(Math.floor(x / size), Math.floor(y / size), Math.floor(z / size)));
    }
  }

  let denominator = 0;
  let numerator = 0;
  for (let y = bottom; y < top; y += 1) {
    for (let x = lo; x < hi; x += 1) {
      for (let z = lo; z < hi; z += 1) {
        if (!inShell((x + 0.5) * size, (y + 0.5) * size, (z + 0.5) * size)) {
          continue;
        }
        denominator += 1;
        if (occupied.has(key(x, y, z))) numerator += 1;
      }
    }
  }
  return denominator ? fraction(numerator, denominator) : untested("no-shell-cells");
}

/** Fraction of terminals within `distance` metres of any colonization
 *  tip. Appended children do not remove a tip from the reference set. */
export function tipClustering(
  skeleton: Skeleton,
  crossover: number,
  terminal: Uint8Array,
  distance: number,
): FillMeasurement {
  const selected = terminals(skeleton, terminal);
  if (!Array.isArray(selected)) return selected;
  if (!Number.isInteger(crossover) || crossover < 1 || crossover > skeleton.nodes.length) {
    return untested("invalid-crossover");
  }
  if (!Number.isFinite(distance) || distance < 0) return untested("invalid-distance");
  const children = new Uint8Array(crossover);
  for (let i = 1; i < crossover; i += 1) {
    const parent = skeleton.nodes[i].parent;
    if (!Number.isInteger(parent) || parent < 0 || parent >= i) {
      return untested("invalid-colonization-topology");
    }
    children[parent] = 1;
  }
  const tips = [];
  for (let i = 1; i < crossover; i += 1) {
    if (children[i] !== 0) continue;
    const position = skeleton.nodes[i].position;
    if (![position.x, position.y, position.z].every(Number.isFinite)) {
      return untested("non-finite-colonization-tip");
    }
    tips.push(position);
  }
  if (!tips.length) return untested("no-colonization-tips");
  let clustered = 0;
  for (const i of selected) {
    const position = skeleton.nodes[i].position;
    for (const tip of tips) {
      if (Math.hypot(position.x - tip.x, position.y - tip.y, position.z - tip.z) <= distance) {
        clustered += 1;
        break;
      }
    }
  }
  return fraction(clustered, selected.length);
}

File: crates/telperion-core/src/colonization/fill.rs
//! Measurements of caller-classified fine wood against the authored crown.
use crate::{
    envelope::{distance_to_profile, Envelope},
    tree::Tree,
    Error, Result,
};
use std::collections::HashSet;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FillMeasurement {
    Tested {
        numerator: usize,
        denominator: usize,
    },
    Untested(&'static str),
}
impl FillMeasurement {
    pub fn fraction(self) -> Option<f64> {
        match self {
            Self::Tested {
                numerator,
                denominator,
            } => Some(numerator as f64 / denominator as f64),
            Self::Untested(_) => None,
        }
    }
}
fn terminals(tree: &Tree, terminal: &[bool]) -> Result<Vec<usize>> {
    tree.validate()?;
    if terminal.len() != tree.nodes.len() {
        return Err(Error::InvalidInput("terminal mask length"));
    }
    Ok(terminal
        .iter()
        .enumerate()
        .filter_map(|(i, &selected)| selected.then_some(i))
        .collect())
}

/// Equal-volume cells, classified at their centres. Sizes are fractions of height
/// and maximum radius respectively; over two million candidate cells is an error.
pub fn shell_occupancy(
    tree: &Tree,
    terminal: &[bool],
    envelope: Envelope,
    cell_size: f64,
    shell_depth: f64,
) -> Result<FillMeasurement> {
    let selected = terminals(tree, terminal)?;
    envelope.validate()?;
    if !cell_size.is_finite() || cell_size <= 0.0 || !shell_depth.is_finite() || shell_depth < 0.0 {
        return Err(Error::InvalidInput("fill resolution or shell depth"));
    }
    if selected.is_empty() {
        return Ok(FillMeasurement::Untested("no-terminals"));
    }
    let size = cell_size * envelope.height;
    let radius = envelope.max_radius();
    let shell = shell_depth * radius;
    if !size.is_finite() || !shell.is_finite() {
        return Err(Error::InvalidInput("fill scale overflow"));
    }
    if size == 0.0 || radius == 0.0 || envelope.crown_base == 1.0 {
        return Ok(FillMeasurement::Untested("no-shell-cells"));
    }
    let base = envelope.height * envelope.crown_base;
    let lo = (-radius / size).floor();
    let hi = (radius / size).ceil();
    let bottom = (base / size).floor();
    let top = (envelope.height / size).ceil();
    let cells = (hi - lo).powi(2) * (top - bottom);
    if ![lo, hi, bottom, top, cells]
        .iter()
        .all(|x| x.is_finite() && x.abs() <= (1_u64 << 53) as f64)
        || cells > 2_000_000.0
    {
        return Err(Error::ResourceLimit("fill grid"));
    }
    let profile = envelope.profile();
    let in_shell = |x: f64, y: f64, z: f64| {
        if y < base || y > envelope.height {
            return false;
        }
        let r = x.hypot(z);
        let slack = envelope.radius_at(y) - r;
        slack >= 0.0 && (slack <= shell || distance_to_profile(&profile, r, y) <= shell)
    };
    let mut occupied = HashSet::new();
    for i in selected {
        let p = tree.nodes[i].position;
        if in_shell(p.x, p.y, p.z) {
            occupied.insert((
                (p.x / size).floor() as i64,
                (p.y / size).floor() as i64,
                (p.z / size).floor() as i64,
            ));
        }
    }
    let mut denominator = 0;
    let mut numerator = 0;
    for y in bottom as i64..top as i64 {
        for x in lo as i64..hi as i64 {
            for z in lo as i64..hi as i64 {
                if in_shell(
                    (x as f64 + 0.5) * size,
                    (y as f64 + 0.5) * size,
                    (z as f64 + 0.5) * size,
                ) {
                    denominator += 1;
                    numerator += usize::from(occupied.contains(&(x, y, z)));
                }
            }
        }
    }
    Ok(if denominator == 0 {
        FillMeasurement::Untested("no-shell-cells")
    } else {
        FillMeasurement::Tested {
            numerator,
            denominator,
        }
    })
}

/// Appended children do not remove a tip from the colonization reference set.
pub fn tip_clustering(tree: &Tree, terminal: &[bool], distance: f64) -> Result<FillMeasurement> {
    let selected = terminals(tree, terminal)?;
    if !distance.is_finite() || distance < 0.0 {
        return Err(Error::InvalidInput("clustering distance"));
    }
    if selected.is_empty() {
        return Ok(FillMeasurement::Untested("no-terminals"));
    }
    let mut children = vec![false; tree.crossover];
    for n in &tree.nodes[..tree.crossover] {
        if let Some(p) = n.parent {
            children[p as usize] = true;
        }
    }
    let tips: Vec<_> = (1..tree.crossover)
        .filter(|&i| !children[i])
        .map(|i| tree.nodes[i].position)
        .collect();
    if tips.is_empty() {
        return Ok(FillMeasurement::Untested("no-colonization-tips"));
    }
    let numerator = selected
        .iter()
        .filter(|&&i| {
            tips.iter()
                .any(|p| p.distance(tree.nodes[i].position) <= distance)
        })
        .count();
    Ok(FillMeasurement::Tested {
        numerator,
        denominator: selected.len(),
    })
}
